be/src/storage/size_tiered_compaction_policy.cpp

```diff
@@ -0,0 +1,277 @@
+
+#include "runtime/current_thread.h"
+#include "storage/compaction_task_factory.h"
+#include "util/defer_op.h"
+#include "util/starrocks_metrics.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace starrocks::vectorized {
+
+SizeTieredCompactionPolicy::SizeTieredCompactionPolicy(Tablet* tablet) : _tablet(tablet) {
+    _max_level_size =
+            config::size_tiered_min_level_size * pow(config::size_tiered_level_multiple, config::size_tiered_level_num);
+}
+
+bool SizeTieredCompactionPolicy::need_compaction(double* score, CompactionType* type) {
+    auto st = _pick_rowsets_to_size_tiered_compact(&_rowsets, &_score);
+
+    if (st.ok()) {
+        if (_rowsets[0]->start_version() == 0) {
+            _compaction_type = BASE_COMPACTION;
+        } else {
+            _compaction_type = CUMULATIVE_COMPACTION;
+        }
+        *score = _score;
+    } else {
+        _compaction_type = INVALID_COMPACTION;
+        *score = 0;
+    }
+    *type = _compaction_type;
+
+    return _compaction_type != INVALID_COMPACTION;
+}
+
+std::shared_ptr<CompactionTask> SizeTieredCompactionPolicy::create_compaction(TabletSharedPtr tablet) {
+    if (_compaction_type != INVALID_COMPACTION) {
+        Version output_version;
+        output_version.first = (*_rowsets.begin())->start_version();
+        output_version.second = (*_rowsets.rbegin())->end_version();
+
+        CompactionTaskFactory factory(output_version, tablet, std::move(_rowsets), _score, _compaction_type);
+        std::shared_ptr<CompactionTask> compaction_task = factory.create_compaction_task();
+        _compaction_type = INVALID_COMPACTION;
+        return compaction_task;
+    }
+    return nullptr;
+}
+
+double SizeTieredCompactionPolicy::_cal_compaction_score(int64_t segment_num, int64_t level_size, int64_t total_size,
+                                                         KeysType keys_type, bool reached_max_version) {
+    double score = 0;
+    if (keys_type == KeysType::DUP_KEYS) {
+        // duplicate keys only has write amplification, so that we use more aggressive size-tiered strategy
+        score = segment_num + ((double)(total_size - level_size) / level_size) * 2;
+    } else {
+        // agg/unique key also has read amplification, segment num occupies a greater weight
+        score = segment_num + (segment_num - 1) * 2 + ((double)(total_size - level_size) / level_size);
+    }
+
+    // level bonus: The lower the level means the smaller the data volume of the compaction, the higher the execution priority
+    int64_t level_bonus = 0;
+    for (int64_t v = level_size; v < _max_level_size && level_bonus <= 7; ++level_bonus) {
+        v = v * config::size_tiered_level_multiple;
+    }
+    score += level_bonus;
+
+    // version limit bonus: The version num of the tablet is about to exceed the limit, we let it perform compaction faster and reduce the version num
+    if (reached_max_version) {
+        score *= 2;
```

Comment:
How about setting the score as a very large value when `reached_max_version==true`?